verifier/ita/client.go

```diff
@@ -0,0 +1,259 @@
+
+	challengeNamePrefix = "ita://"
+)
+
+var regionalURLs map[string]string = map[string]string{
+	"US": "https://api.trustauthority.intel.com",
+	"EU": "https://api.eu.trustauthority.intel.com",
+}
+
+type client struct {
+	inner  *http.Client
+	apiURL string
+	apiKey string
+}
+
+func urlFromRegion(region string) (string, error) {
+	if region == "" {
+		return "", errors.New("API region required to initialize ITA client")
+	}
+
+	url, ok := regionalURLs[strings.ToUpper(region)]
+	if !ok {
+		// Create list of allowed regions.
+		keys := []string{}
+		for k := range regionalURLs {
+			keys = append(keys, k)
+		}
+		return "", fmt.Errorf("unsupported region %v, expect one of %v", region, keys)
+	}
+
+	return url, nil
+}
+
+func NewClient(region string, key string) (verifier.Client, error) {
+	url, err := urlFromRegion(region)
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		inner: &http.Client{
+			Transport: &http.Transport{
+				// https://github.com/intel/trustauthority-client-for-go/blob/main/go-connector/token.go#L130.
+				TLSClientConfig: &tls.Config{
+					CipherSuites: []uint16{
+						tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+						tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+					},
+					InsecureSkipVerify: false,
+					MinVersion:         tls.VersionTLS12,
+				},
+				Proxy: http.ProxyFromEnvironment,
+			},
+		},
+		apiURL: url,
+		apiKey: key,
+	}, nil
+}
+
+// Confirm that client implements verifier.Client interface.
+var _ verifier.Client = (*client)(nil)
+
+type itaNonce struct {
+	Val       []byte `json:"val"`
+	Iat       []byte `json:"iat"`
+	Signature []byte `json:"signature"`
+}
+
+// The ITA evidence nonce is a concatenation+hash of Val and Iat. See references below:
+// https://github.com/intel/trustauthority-client-for-go/blob/main/go-connector/attest.go#L22
+// https://github.com/intel/trustauthority-client-for-go/blob/main/go-tdx/tdx_adapter.go#L37
+func createHashedNonce(nonce *itaNonce) ([]byte, error) {
+	hash := sha512.New()
+	_, err := hash.Write(append(nonce.Val, nonce.Iat...))
+	if err != nil {
+		return nil, fmt.Errorf("error hashing ITA nonce: %v", err)
+	}
+
+	return hash.Sum(nil), err
+}
+
+func (c *client) CreateChallenge(_ context.Context) (*verifier.Challenge, error) {
+	url := c.apiURL + nonceEndpoint
+
+	headers := map[string]string{
+		apiKeyHeader: c.apiKey,
+		acceptHeader: applicationJSON,
+	}
+
+	resp := &itaNonce{}
+	if err := c.doHTTPRequest(http.MethodGet, url, nil, headers, &resp); err != nil {
+		return nil, err
+	}
+
+	nonce, err := createHashedNonce(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return &verifier.Challenge{
+		Name:      challengeNamePrefix + string(resp.Val),
+		Nonce:     nonce,
+		Val:       resp.Val,
+		Iat:       resp.Iat,
+		Signature: resp.Signature,
+	}, nil
+}
+
+func convertRequestToTokenRequest(request verifier.VerifyAttestationRequest) tokenRequest {
+	// Trim trailing 0xFF bytes from CCEL Data.
+	data := request.TDCCELAttestation.CcelData
```

Comment:
Why is this happening here and not in the agent?

Reply:
The request to trim the data came from Intel so I put it in the ITA-specific code.